### PR TITLE
better sharding

### PR DESCRIPTION
**Pull request by bheni**
Previously shards were cut whenever a threshold was reached.  With a threshold of N it was possible for N+1 items to be split into a shard with N items, and a shard with 1 item.  This happens much more frequently when merge commits have very few rows in them and only a handful of items are added to a shard near the threshold.  What makes it worse is that it's likely to happen again and again in bounties where merge commits are small resulting in a big shard, and many small shards.

The new algorithm writes items to a queue. We need a queue because we need to maintain the order that items are written. If more items are in the queue than are in the current shard being written to we move items from the queue to the current shard.  At the point that the current shard reaches the threshold the queue will be of equal size. The current shard is closed, a new shard is opened, and half the items from the queue are moved to the new shard.  Work continues until all items have been added.  At this point we look at the number of rows that have been written to the current shard and the number that have been queued.  If their sum M exceeds our threshold of N we write two shards with M/2 shards.  If their sum is less than the threshold all queued items are written to the current shard and it is closed.

### CODE REVIEW

**Review by coffeegoddd — APPROVED** [review]
LGTM!

File: go/payments/pkg/cellwise/shard_manager.go [hunk 1]
```diff
@@ -66,53 +128,85 @@ func (sm *shardManager) getShards() []AttributionShard {
...
 	}
 
-	// open a new shard if not actively writing a shard
-	if sm.currStore == nil {
-		err := sm.openNewShard(ctx, key)
+	sm.aQ.Push(key, raVal, ra)
+	nextKey := sm.aQ.PeekKey()
+
+	// check and shard if beyond the configured number of rows per shard
+	if sm.rowsBufferred == sm.shardParams.RowsPerShard {
```
Comment (katiemcculloch) [inline go/payments/pkg/cellwise/shard_manager.go:144]: Would it be safer to do `sm.rowsBufferred >= sm.shardParams.RowsPerShard` ?

**Review by katiemcculloch — APPROVED** [review]

File: go/payments/pkg/cellwise/shard_manager.go (hunk 1 above)
Comment (bheni) [reply go/payments/pkg/cellwise/shard_manager.go:144]: Yes. But in practice it's the same.  This is checked every time a row is added, so it shouldn't matter... but I made the change.

File: go/payments/pkg/cellwise/shard_manager.go [hunk 2]
```diff
@@ -66,53 +128,85 @@ func (sm *shardManager) getShards() []AttributionShard {
 
 // add an attributed row to the current shard being built
 func (sm *shardManager) addRowAtt(ctx context.Context, key types.Value, ra rowAtt, raVal types.Value) error {
-	// check and shard if beyond the configured number of rows per shard
-	if sm.rowsInCurrShard >= sm.shardParams.RowsPerShard {
-		err := sm.closeCurrentShard(ctx, key)
+	// if necessary encode the row attribution data as a noms value
+	if raVal == nil {
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/payments/pkg/cellwise/shard_manager.go:132]: raVal is only Non-nil in cases where the previous attribution for the row is being used unchanged.  This happens in cases where a row is unchanged in a commit so the attribution stays the same.

**Pull request merged**